File: src/adapters/voice/protocol.ts
// Voice WebSocket protocol types

// --- Client -> Server ---

export interface TextMessage {
  type: 'text';
  requestId: string;
  text: string;
  tts?: boolean;
}

export interface AudioStartMessage {
  type: 'audio_start';
  requestId: string;
  format: { mimeType: string; sampleRate?: number; channels?: number; encoding?: string };
  tts?: boolean;
}

export interface AudioChunkMessage {
  type: 'audio_chunk';
  requestId: string;
  data: string; // base64-encoded audio bytes
}

export interface AudioEndMessage {
  type: 'audio_end';
  requestId: string;
}

export interface CancelMessage {
  type: 'cancel';
  requestId: string;
}

export interface PingMessage {
  type: 'ping';
}

export interface AuthMessage {
  type: 'auth';
  token: string;
}

export type VoiceClientMessage =
  | TextMessage
  | AudioStartMessage
  | AudioChunkMessage
  | AudioEndMessage
  | CancelMessage
  | PingMessage
  | AuthMessage;

// --- Server -> Client ---

export interface StatusMessage {
  type: 'status';
  requestId: string;
  status: string;
  toolName?: string;
  keyArg?: string;
  phase?: string;
  description?: string;
  usage?: { toolUses: number; tokens: number; durationMs: number };
}

export interface ResponseTextMessage {
  type: 'response_text';
  requestId: string;
  text: string;
  final: boolean;
}

export interface TranscriptMessage {
  type: 'transcript';
  requestId: string;
  text: string;
  final: boolean;
}

export interface ResponseAudioMessage {
  type: 'response_audio';
  requestId: string;
  data: string; // base64-encoded audio bytes
  encoding: string; // e.g. 'linear16'
  sampleRate: number; // e.g. 24000
}

export interface ResponseAudioEndMessage {
  type: 'response_audio_end';
  requestId: string;
}

export interface ErrorMessage {
  type: 'error';
  requestId: string | null;
  message: string;
}

export interface ChannelInfoMessage {
  type: 'channel_info';
  channelId: string;
  channelName: string;
  repo: string | null;
  model: string;
}

export interface PongMessage {
  type: 'pong';
}

export type VoiceServerMessage =
  | StatusMessage
  | ResponseTextMessage
  | ResponseAudioMessage
  | ResponseAudioEndMessage
  | TranscriptMessage
  | ErrorMessage
  | ChannelInfoMessage
  | PongMessage;

const CLIENT_MESSAGE_TYPES = new Set([
  'text',
  'audio_start',
  'audio_chunk',
  'audio_end',
  'cancel',
  'ping',
  'auth',
]);

/**
 * Parse and validate a raw WebSocket message into a typed client message.
 * Throws on invalid input.
 */
export function parseClientMessage(raw: string): VoiceClientMessage {
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    throw new Error('Invalid JSON');
  }

  if (typeof parsed !== 'object' || parsed === null) {
    throw new Error('Message must be a JSON object');
  }

  const msg = parsed as Record<string, unknown>;

  if (typeof msg.type !== 'string' || !CLIENT_MESSAGE_TYPES.has(msg.type)) {
    throw new Error(`Unknown or missing message type: ${String(msg.type)}`);
  }

  if (msg.type === 'ping') {
    return { type: 'ping' };
  }

  if (msg.type === 'auth') {
    if (typeof msg.token !== 'string' || msg.token.length === 0) {
      throw new Error('Missing or empty token field');
    }
    return { type: 'auth', token: msg.token };
  }

  if (typeof msg.requestId !== 'string' || msg.requestId.length === 0) {
    throw new Error('Missing or empty requestId');
  }

  if (msg.type === 'text') {
    if (typeof msg.text !== 'string') {
      throw new Error('Missing text field');
    }
    return {
      type: 'text',
      requestId: msg.requestId,
      text: msg.text,
      ...(typeof msg.tts === 'boolean' ? { tts: msg.tts } : {}),
    };
  }

  if (msg.type === 'audio_start') {
    if (typeof msg.format !== 'object' || msg.format === null) {
      throw new Error('Missing format object');
    }
    const fmt = msg.format as Record<string, unknown>;
    if (typeof fmt.mimeType !== 'string' || fmt.mimeType.length === 0) {
      throw new Error('Missing or empty format.mimeType');
    }
    return {
      type: 'audio_start',
      requestId: msg.requestId,
      format: {
        mimeType: fmt.mimeType,
        ...(typeof fmt.sampleRate === 'number' ? { sampleRate: fmt.sampleRate } : {}),
        ...(typeof fmt.channels === 'number' ? { channels: fmt.channels } : {}),
        ...(typeof fmt.encoding === 'string' ? { encoding: fmt.encoding } : {}),
      },
      ...(typeof msg.tts === 'boolean' ? { tts: msg.tts } : {}),
    };
  }

  if (msg.type === 'audio_chunk') {
    if (typeof msg.data !== 'string' || msg.data.length === 0) {
      throw new Error('Missing or empty data field');
    }
    return { type: 'audio_chunk', requestId: msg.requestId, data: msg.data };
  }

  if (msg.type === 'audio_end') {
    return { type: 'audio_end', requestId: msg.requestId };
  }

  // cancel
  return { type: 'cancel', requestId: msg.requestId };
}

/**
 * Serialize a server message to JSON string.
 */
export function serializeServerMessage(msg: VoiceServerMessage): string {
  return JSON.stringify(msg);
}

File: android/app/src/main/kotlin/com/claudeway/network/Protocol.kt
package com.claudeway.network

import org.json.JSONObject

// --- Audio format descriptor (matches server AudioFormat) ---

data class AudioFormat(
    val mimeType: String,
    val sampleRate: Int? = null,
    val channels: Int? = null,
    val encoding: String? = null,
)

// --- Client -> Server messages ---

sealed interface ClientMessage {
    fun toJson(): String
}

data class TextMessage(
    val requestId: String,
    val text: String,
) : ClientMessage {
    override fun toJson(): String = JSONObject().apply {
        put("type", "text")
        put("requestId", requestId)
        put("text", text)
    }.toString()
}

data class AudioStartMessage(
    val requestId: String,
    val format: AudioFormat,
) : ClientMessage {
    override fun toJson(): String = JSONObject().apply {
        put("type", "audio_start")
        put("requestId", requestId)
        put("format", JSONObject().apply {
            put("mimeType", format.mimeType)
            format.sampleRate?.let { put("sampleRate", it) }
            format.channels?.let { put("channels", it) }
            format.encoding?.let { put("encoding", it) }
        })
    }.toString()
}

data class AudioChunkMessage(
    val requestId: String,
    val data: String, // base64-encoded audio bytes
) : ClientMessage {
    override fun toJson(): String = JSONObject().apply {
        put("type", "audio_chunk")
        put("requestId", requestId)
        put("data", data)
    }.toString()
}

data class AudioEndMessage(
    val requestId: String,
) : ClientMessage {
    override fun toJson(): String = JSONObject().apply {
        put("type", "audio_end")
        put("requestId", requestId)
    }.toString()
}

data class CancelMessage(
    val requestId: String,
) : ClientMessage {
    override fun toJson(): String = JSONObject().apply {
        put("type", "cancel")
        put("requestId", requestId)
    }.toString()
}

class PingMessage : ClientMessage {
    override fun toJson(): String = """{"type":"ping"}"""
}

// --- Server -> Client messages ---

sealed interface ServerMessage

data class ToolUsage(
    val toolUses: Int = 0,
    val tokens: Int = 0,
    val durationMs: Long = 0,
)

data class StatusServerMessage(
    val requestId: String,
    val status: String,
    val toolName: String? = null,
    val keyArg: String? = null,
    val phase: String? = null,
    val description: String? = null,
    val usage: ToolUsage? = null,
) : ServerMessage

data class TranscriptServerMessage(
    val requestId: String,
    val text: String,
    val isFinal: Boolean,
) : ServerMessage

data class ResponseTextServerMessage(
    val requestId: String,
    val text: String,
    val isFinal: Boolean,
) : ServerMessage

data class ResponseAudioServerMessage(
    val requestId: String,
    val data: String, // base64-encoded audio bytes
    val encoding: String, // e.g. "linear16"
    val sampleRate: Int, // e.g. 24000
) : ServerMessage

data class ResponseAudioEndServerMessage(
    val requestId: String,
) : ServerMessage

data class ErrorServerMessage(
    val requestId: String?,
    val message: String,
) : ServerMessage

data class ChannelInfoServerMessage(
    val channelId: String,
    val channelName: String,
    val repo: String?,
    val model: String,
) : ServerMessage

class PongServerMessage : ServerMessage

// --- Serialization helpers ---

object ProtocolAdapters {
    /** Parse a server JSON message by reading the "type" field and dispatching. */
    fun parseServerMessage(json: String): ServerMessage? {
        return try {
            val obj = JSONObject(json)
            when (obj.optString("type")) {
                "status" -> {
                    val usageObj = obj.optJSONObject("usage")
                    val usage = if (usageObj != null) {
                        ToolUsage(
                            toolUses = usageObj.optInt("toolUses", 0),
                            tokens = usageObj.optInt("tokens", 0),
                            durationMs = usageObj.optLong("durationMs", 0),
                        )
                    } else null
                    StatusServerMessage(
                        requestId = obj.getString("requestId"),
                        status = obj.getString("status"),
                        toolName = obj.optStringOrNull("toolName"),
                        keyArg = obj.optStringOrNull("keyArg"),
                        phase = obj.optStringOrNull("phase"),
                        description = obj.optStringOrNull("description"),
                        usage = usage,
                    )
                }
                "transcript" -> TranscriptServerMessage(
                    requestId = obj.getString("requestId"),
                    text = obj.getString("text"),
                    isFinal = obj.getBoolean("final"),
                )
                "response_text" -> ResponseTextServerMessage(
                    requestId = obj.getString("requestId"),
                    text = obj.getString("text"),
                    isFinal = obj.getBoolean("final"),
                )
                "response_audio" -> ResponseAudioServerMessage(
                    requestId = obj.getString("requestId"),
                    data = obj.getString("data"),
                    encoding = obj.getString("encoding"),
                    sampleRate = obj.getInt("sampleRate"),
                )
                "response_audio_end" -> ResponseAudioEndServerMessage(
                    requestId = obj.getString("requestId"),
                )
                "error" -> ErrorServerMessage(
                    requestId = if (obj.isNull("requestId")) null else obj.getString("requestId"),
                    message = obj.getString("message"),
                )
                "channel_info" -> ChannelInfoServerMessage(
                    channelId = obj.getString("channelId"),
                    channelName = obj.getString("channelName"),
                    repo = if (obj.isNull("repo")) null else obj.optString("repo", null),
                    model = obj.getString("model"),
                )
                "pong" -> PongServerMessage()
                else -> null
            }
        } catch (_: Exception) {
            null
        }
    }

    /** Serialize a client message to JSON. */
    fun serializeClientMessage(msg: ClientMessage): String = msg.toJson()

    /** Helper: returns null instead of empty string for missing JSON keys. */
    private fun JSONObject.optStringOrNull(key: String): String? {
        return if (has(key) && !isNull(key)) getString(key) else null
    }
}
